src/surroundings_menu.cpp

```diff
@@ -0,0 +1,1160 @@
+
+    }
+    ImGui::TableNextRow();
+}
+
+void surroundings_menu::draw_item_tab()
+{
+    ImGuiTabItemFlags_ flags = ImGuiTabItemFlags_None;
+    if( switch_tab == surroundings_menu_tab_enum::items ) {
+        flags = ImGuiTabItemFlags_SetSelected;
+        switch_tab = surroundings_menu_tab_enum::num_tabs;
+    }
+    if( ImGui::BeginTabItem( item_data.title.c_str(), nullptr, flags ) ) {
+        selected_tab = surroundings_menu_tab_enum::items;
+        if( ImGui::BeginTable( "items", 2, ImGuiTableFlags_ScrollY,
+                               ImVec2( 0.0f, str_height_to_pixels( TERMY - info_height - 3 ) ) ) ) {
+            ImGui::TableSetupColumn( "it_name", ImGuiTableColumnFlags_NoDirectResize_,
+                                     str_width_to_pixels( width - dist_width - 3 ) );
+            ImGui::TableSetupColumn( "it_distance", ImGuiTableColumnFlags_NoDirectResize_,
+                                     str_width_to_pixels( dist_width ) );
+            if( !item_data.selected_entry ||
+                std::find( item_data.filtered_list.begin(), item_data.filtered_list.end(),
+                           item_data.selected_entry ) == item_data.filtered_list.end() ) {
+                item_data.selected_entry = item_data.filtered_list.front();
+            }
+            std::string last_category;
+            int entry_no = 0;
+            for( map_entity_stack<item> *it : item_data.filtered_list ) {
+                if( item_data.draw_categories ) {
+                    draw_category_separator( it->get_category(), last_category, 0 );
+                }
+                bool is_selected = it == item_data.selected_entry;
+                ImGui::TableNextColumn();
+                ImGui::PushID( entry_no++ );
+                ImGui::Selectable( "", &is_selected,
+                                   ImGuiSelectableFlags_AllowItemOverlap | ImGuiSelectableFlags_SpanAllColumns, ImVec2( 0, 0 ) );
+                if( is_selected ) {
+                    if( auto_scroll ) {
+                        ImGui::SetScrollHereY();
+                        auto_scroll = false;
+                    }
+                    item_data.selected_entry = it;
+                }
+                ImGui::SameLine( 0, 0 );
+                ImGui::PopID();
+                const item *itm = it->get_selected_entity();
+                std::string text;
+                if( it->entities.size() > 1 ) {
+                    text = string_format( "[%d/%d] (%d) ", it->get_selected_index() + 1, it->entities.size(),
+                                          it->totalcount );
+                }
+
+                const int count = it->get_selected_count();
+                if( count > 1 ) {
+                    text += string_format( "%d ", count );
+                }
+
+                text += itm->display_name( count );
+                nc_color color = it->get_selected_entity()->color_in_inventory();
+                cataimgui::draw_colored_text( text, color );
```

Comment:
Instead of allocating a string and appending a bunch of things to it (which can cause reallocation) before finally drawing the string, it would be better to simply call `ImGui::Text`, `ImGui::TextUnformatted`, or `cataimgui::draw_colored_text` for each element individually. You can arrange multiple items on the same line by calling `ImGui::SameLine(0, 0)`. Remember that all of this happens on every frame, so it’s best to avoid allocation.

Reply:
I'm not a fan of splitting up the drawing like that. I opted for caching the strings instead.

Reply:
Ok. But note that `ImGui::Text` takes a format string and works without allocating (it reuses a fixed-size temporary buffer for every call).